servo-media/src/backends/gstreamer/audio_decoder.rs

```diff
@@ -0,0 +1,195 @@
+    }
+}
+
+impl AudioDecoder for GStreamerAudioDecoder {
+    fn decode(&self, data: Vec<u8>, callbacks: AudioDecoderCallbacks) {
+        let pipeline = gst::Pipeline::new(None);
+        let callbacks = Arc::new(callbacks);
+
+        let appsrc = match gst::ElementFactory::make("appsrc", None) {
+            Some(appsrc) => appsrc,
+            None => return callbacks.error(),
+        };
+
+        let decodebin = match gst::ElementFactory::make("decodebin", None) {
+            Some(decodebin) => decodebin,
+            None => return callbacks.error(),
+        };
+
+        // decodebin uses something called a "sometimes-pad", which is basically
+        // a pad that will show up when a certain condition is met,
+        // in decodebins case that is media being decoded
+        if pipeline.add_many(&[&appsrc, &decodebin]).is_err() {
+            return callbacks.error();
+        }
+
+        if gst::Element::link_many(&[&appsrc, &decodebin]).is_err() {
+            return callbacks.error();
+        }
+
+        let appsrc = match appsrc.downcast::<AppSrc>() {
+            Ok(appsrc) => appsrc,
+            Err(_) => {
+                return callbacks.error();
+            }
+        };
+
+        let pipeline_ = pipeline.clone();
+        let callbacks_ = callbacks.clone();
+        decodebin.connect_pad_added(move |_, src_pad| {
+            // We only want a sink per audio stream.
+            if src_pad.is_linked() {
+                return;
+            }
+
+            let pipeline = &pipeline_;
+            let callbacks = &callbacks_;
+
+            let (is_audio, _) = {
+                let media_type = src_pad.get_current_caps().and_then(|caps| {
+                    caps.get_structure(0).map(|s| {
+                        let name = s.get_name();
+                        (name.starts_with("audio/"), name.starts_with("video/"))
+                    })
+                });
+
+                match media_type {
+                    None => {
+                        eprintln!("Failed to get media type from pad {}", src_pad.get_name());
+                        return callbacks.error();
+                    }
+                    Some(media_type) => media_type,
+                }
+            };
+
+            if !is_audio {
+                return callbacks.error();
+            }
+
+            let insert_sink = || -> Result<(), ()> {
+                let queue = gst::ElementFactory::make("queue", None).ok_or(())?;
+                let convert = gst::ElementFactory::make("audioconvert", None).ok_or(())?;
+                let resample = gst::ElementFactory::make("audioresample", None).ok_or(())?;
+                let sink = gst::ElementFactory::make("appsink", None).ok_or(())?;
+                let appsink = sink.clone().dynamic_cast::<AppSink>().map_err(|_| ())?;
+
+                // XXX Issue #18 channels support.
+                let audio_info = gst_audio::AudioInfo::new(gst_audio::AUDIO_FORMAT_F32, 44100, 1)
```

Comment:
If 44100 Hz are not required, 48000 might be nicer... or can the sample rate even be signalled somewhere via webaudio so you wouldn't have to constrain it here?

Reply:
Yeah, the sample rate should probably be given by the AudioContext. I'll write a patch for this.

Reply:
You then want to set the caps similar to this (and possibly later also allow a different number of channels): https://github.com/sdroege/gstreamer-rs/blob/db3fe694154c697afdaf3efb6ec65332546942e0/examples/src/bin/appsink.rs#L54-L62